ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos/VRF.hs

```diff
@@ -0,0 +1,123 @@
+import           Cardano.Ledger.BaseTypes (Nonce (NeutralNonce, Nonce))
+import           Cardano.Ledger.Crypto (Crypto (HASH, VRF))
+import           Cardano.Ledger.Serialization (runByteBuilder)
+import           Cardano.Ledger.Slot (SlotNo (SlotNo))
+import qualified Data.ByteString.Builder as BS
+import qualified Data.ByteString.Builder.Extra as BS
+import           GHC.Generics (Generic)
+import           NoThunks.Class (NoThunks)
+import           Numeric.Natural (Natural)
+
+-- | Input to the verifiable random function. Consists of the hash of the slot
+-- and the epoch nonce.
+newtype InputVRF = InputVRF {unUnifiedVRF :: Hash Blake2b_256 InputVRF}
+  deriving (Eq, Ord, Show, Generic)
+  deriving newtype (NoThunks, ToCBOR)
+
+instance SignableRepresentation InputVRF where
+  getSignableRepresentation (InputVRF x) = hashToBytes x
+
+-- | Construct a unified VRF value
+mkInputVRF ::
+  SlotNo ->
+  -- | Epoch nonce
+  Nonce ->
+  InputVRF
+mkInputVRF (SlotNo slot) eNonce =
+  InputVRF
+    . Hash.castHash
+    . Hash.hashWith id
+    . runByteBuilder (8 + 32)
+    $ BS.word64BE slot
+      <> ( case eNonce of
+             NeutralNonce -> mempty
+             Nonce h      -> BS.byteStringCopy (Hash.hashToBytes h)
+         )
+
+-- | Indicate the usage of the VRF result.
+data VRFUsage
+  = -- | The VRF value will be used to establish whether the issuing node is
+    -- indeed a leader for this slot.
+    VRFLeader
+  | -- | The VRF value will be used to contribute to the evolving nonce.
+    VRFNonce
+
+-- | Singleton VRF usage
+data SVRFUsage a where
+  SVRFLeader :: SVRFUsage VRFLeader
+  SVRFNonce :: SVRFUsage VRFNonce
+
+class VRFUsageTypeRep t where
+  vuTypeRep :: SVRFUsage t
+
+instance VRFUsageTypeRep VRFLeader where
+  vuTypeRep = SVRFLeader
+
+instance VRFUsageTypeRep VRFNonce where
+  vuTypeRep = SVRFNonce
+
+-- | Indicate the result of the VRF evaluation.
+data VRFResult (v :: VRFUsage)
+
+-- | Compute a hash of the unified VRF output appropriate to its usage.
+hashVRF ::
+  forall (v :: VRFUsage) c proxy.
+  (Crypto c, VRFUsageTypeRep v) =>
+  proxy c ->
+  OutputVRF (VRF c) ->
+  Hash (HASH c) (VRFResult v)
+hashVRF _ certVRF =
+  let vrfOutputAsBytes = getOutputVRFBytes certVRF
+   in case vuTypeRep @v of
+        SVRFLeader -> castHash $ hashWith id $ "L" <> vrfOutputAsBytes
+        SVRFNonce  -> castHash $ hashWith id $ "N" <> vrfOutputAsBytes
+
+-- | Simulate a VRF output to be used for leader checks from the relevant hash.
+vrfLeaderCheckValue ::
+  forall c proxy.
+  Crypto c =>
+  proxy c ->
+  OutputVRF (VRF c) ->
+  Natural
+vrfLeaderCheckValue p =
+  bytesToNatural . hashToBytes . hashVRF @VRFLeader p
+
+-- | Simulate a VRF output to be used for the evolving nonce.
+vrfNonceValue ::
+  forall c proxy.
+  Crypto c =>
+  proxy c ->
+  OutputVRF (VRF c) ->
+  Nonce
+vrfNonceValue p =
+  Nonce . castHash . hashWith id . hashToBytes . hashVRF @VRFNonce p
```

Comment:
It's done to be consistent with existing behaviour.

The first hash is how we do range extension as per the VRF paper. The second hash is how we generate a nonce value from a VRF output. It's just somewhat confusing that now, the "VRF output" is actually a hash output.

We cannot drop this in any case, since the VRF hash is crypto-dependent, whereas the Nonce hash is of a fixed type `Blake2b_256`.

Reply:
I'll add a comment to explain this.

Reply:
I'm not totally convinced: the nonce type uses a fixed hash type, but afaik that's only because we didn't bother to parametrise it. In practice these are the same hashes. I expect if we asked the researchers they'd say it's fine to hash once now.

But I get that the types here mean we cannot assume the nonce hash type matches the other hash. Perhaps we can take the same cheeky approach and have the VRF output hashing be a fixed hash type too.

But if we keep it like this, we could do slightly better to match the style before where we had `mkNonceFromOutputVRF`. So now we're not starting from an `OutputVRF` but from another hash. But we could still use a helper like this. So I'd suggest both code structure and comment. It is quite confusing otherwise.